### PR TITLE
Add L1 objective function

### PR DESCRIPTION
**Pull request by henry0312**
- [x] Merge and Rebase #174

### CODE REVIEW

**Comment by guolinke** [comment]
It seems this PR have the changes of #174.
I will review this after merged the #174

**Comment by guolinke** [comment]
@henry0312 , can you solve these conflict first ?

**Comment by henry0312** [comment]
@guolinke solved the conflicts

File: src/objective/regression_objective.hpp [hunk 1]
```diff
@@ -51,6 +51,60 @@ class RegressionL2loss: public ObjectiveFunction {
...
+    } else {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+        if (diff >= 0.0) {
+          gradients[i] = weights_[i];
+        } else {
+          gradients[i] = -weights_[i];
+        }
+        hessians[i] = 0.0;
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: all hessians are zero? 
The output of leaf is sum_gradients/sum_hessians. if all hessians are zero, I think it cannot work. 
BTW, it seems some of hessians are zero in huber loss as well. Is this correct?

Comment (henry0312) [inline src/objective/regression_objective.hpp]: > all hessians are zero?

yes, the absolute error is `L(y) = |y - t|`, where `y` means outputs of models and `t` means true targets.
Then, L'(y) = 1.0 or -1.0, and L''(y) = 0.

> if all hessians are zero, I think it cannot work.

oh...😢

> BTW, it seems some of hessians are zero in huber loss as well. Is this correct?

I don't think it is wrong.

Comment (guolinke) [inline src/objective/regression_objective.hpp]: maybe use an approximate function for L1?.
e.g. (y-t)^((k+1)/k) ?

Comment (henry0312) [inline src/objective/regression_objective.hpp]: That sounds good!
Is the `k` is a parameter for loss function, right?

Comment (henry0312) [inline src/objective/regression_objective.hpp]: Also, do I need to fix huber loss?

Comment (guolinke) [inline src/objective/regression_objective.hpp]: yes. But I just give an example, maybe it exists a better way to approx it.

Comment (guolinke) [inline src/objective/regression_objective.hpp]: for huber loss, just keep it now. We can fix it after solved this.

Comment (guolinke) [inline src/objective/regression_objective.hpp]: actually, I think you also can let all hessians to 1. If all hessians == 1, it means doesn't use second order gradients. 
also, you can refer to the implementation of sklearn:
https://github.com/scikit-learn/scikit-learn/blob/14031f6/sklearn/ensemble/gradient_boosting.py#L305-L328

Comment (henry0312) [inline src/objective/regression_objective.hpp]: A friend of mine who majors in theoretical physics tells me that it may be good to approximate hessians with Gaussian function.
Gaussian function is often used to approximate [ultrashort pulse](https://en.wikipedia.org/wiki/Ultrashort_pulse "Ultrashort pulse - Wikipedia").

Which do you think it is better to use `all hessians == 1` or Gaussian function.

Comment (guolinke) [inline src/objective/regression_objective.hpp]: @henry0312 , I think we can keep them both, and use some parameters to switch them.

Comment (henry0312) [inline src/objective/regression_objective.hpp]: ok, I'll try.

**Comment by henry0312** [comment]
@guolinke I fixed hessians. 
Some exprimetns shows me that approximation with Gaussian function is better, and `all hessians == 1` often doesn't work well.

**Comment by henry0312** [comment]
In addition, performances with huber loss gets better!
Thank you for your review, https://github.com/Microsoft/LightGBM/pull/175#discussion_r95103293.
